docs: edit event-binding doc copy and headers (#38903)
angular/angular: aio/content/guide/event-binding.md

# Event binding

> Event binding allows you to listen for and respond to user actions such as keystrokes, mouse movements, clicks, and touches.

Reviewer: I just really like this sentence.

<div class="alert is-helpful">

See the <live-example></live-example> for a working example containing the code snippets in this guide.

</div>

> ## Syntax `(event)`

Reviewer: I'm not sure what `(event)` is doing here. Maybe:

```suggestion
## Binding to events
```

I think this is the title I used in my user-input PR: [38961](https://github.com/angular/angular/pull/38961).

> Angular event binding syntax consists of a target event name within parentheses to the left of an equal sign, and a quoted template statement to the right.

Reviewer: ```suggestion
To bind to an event you use the Angular event binding syntax. This syntax consists of a target event name within parentheses to the left of an equal sign, and a quoted template statement to the right.
```

This rewrite puts the focus on the user task, instead of the Angular feature.

In the following example, the target event name is `click` and the template statement is `onSave()`.

<code-example language="html" header="Event binding syntax">
&lt;button (click)="onSave()"&gt;Save&lt;button&gt;
</code-example>

The event binding listens for the button's click events and calls the component's `onSave()` method whenever a click occurs.

<div class="lightbox">
  <img src='generated/images/guide/template-syntax/syntax-diagram.svg' alt="Syntax diagram">
</div>

> ## Template statements have side effects

Reviewer: ```suggestion
## Template statements and side effects
```
I'm not sure why, but headings normally aren't sentences. I'll see if I can find any information regarding why.

Author: That reminds me of a book about the evolution of language and comparative grammar, "The Loom of Language", that touches on the unique grammar of newspaper headlines. We used to have chapter titles like, "Of martial constitutions or why divers vapors and rising dry heat smelt the humors" that eventually evolved to article titles with wording like, "So-and-so is at it again". I bet it's that--something to do with the 15th century. 😀

> Template statements have side effects.

Reviewer: Always? Or just sometimes?

Author: The current doc has "usually": https://angular.io/guide/event-binding#template-statements-have-side-effects

Maybe an engineer could clarify for us.

Reviewer: I think I'm okay with something like "Often, template statements have side effects. For example..."

Also, did you update the heading? It still reads "Template statements **have** side effects", and I think I prefer **and** instead here.

Reviewer 2: It took me a while to figure out why talking about side-effects here feels weird. I've realized that it's not really the right term to describe what's happening. A "side-effect" describes something that happens incidentally as part of some other action. E.g. in medicine, a side-effect is something that happens in addition to the main effect of a substance.

In the case of event handlers, the thing you're doing is the main effect. If you're button click hander is called `saveProfile()`, you wouldn't say that saving the profile is a "side-effect" of the event happening- it's just the effect. 

Talking about side-effects makes sense for property bindings and text interpolation because, in those APIs, the main "effect" is reading a value. In programming, you generally never want the act of _reading_ a value to mutate the application state. To say it another way, reads should be "side-effect free". But with event handling, it's not a read, so talking about side-effects doesn't really make sense in the same way.

Author: Oh, @aikidave I forgot to push up! Regarding what @jelbourn was saying, what if we delete this section? It's very small and what is explaining is covered elsewhere in the same document.

For example, a `deleteItem()` method has the side effect of deleting an item.

Deleting an item updates the model, and can trigger other changes such as queries or saving to a remote server.
These changes propagate through the system and display in the view.

> ## How event binding works

Reviewer: As I read this section, it seems very conceptual. I recommend removing it so this topic can focus on helping users get stuff done. Thoughts?

Author: Works for me. Should we move it to the Conceptual Reference section as a new document or are you thinking of deleting it?

Reviewer: I suggest one of two courses of action:

1. In this PR, move the conceptual content to a new topic and put it under the Conceptual Reference section.
2. Same as #1, but instead of doing it this PR, create an issue and then address it in a separate PR.

Both options are fine, it just depends on what makes the most sense to you.

Reviewer 2: Honestly, this whole section could probably just be deleted. I don't think it's providing any value for any particular audience. For people new to the framework, they don't need to care how Angular sets up the event binding. At the other end of the spectrum, more expert users would want to know more about Angular's event handling than this outlines (e.g. Angular always adds event listeners to elements directly instead of using event delegation).

In an event binding, Angular configures an event handler for the target event.
You can use event binding  with your own custom events.

When the component or directive raises the event, the handler executes the template statement.
The template statement performs an action in response to the event.

> ### The `$event` object

Reviewer: I suggest we reposition this content to focus on a user task, rather than the Angular feature. Something like:

```suggestion
### Handling events

A common way to handle events is to pass the event object, `$event`, to the method handling the event. The `$event` object often contains information the method needs, such as a user's name or an image URL.
```
etc.

(Note that I won't rewrite the section here. I'm just offering some thoughts.

An event binding conveys information about the event in the `$event` object.
This information can include data values such as an event object, string, or number.

The target event determines the shape of the `$event` object.
If the target event is a native DOM element event, then `$event` is a [DOM event object](https://developer.mozilla.org/en-US/docs/Web/Events), with properties such as `target` and `target.value`.

In the following example the code sets the `<input>` `value` property by binding to the `name` property.

<code-example path="event-binding/src/app/app.component.html" region="event-binding-3" header="src/app/app.component.html"></code-example>

> To listen for changes to the value, the code binds to the `input` event of the `<input>` element.
> When the user makes changes, the component raises the `input` event.
> Next, the binding executes the statement within a context that includes the DOM event object, `$event`.
> 
> To update the `name` property, Angular retreives the changed text by following the path `$event.target.value`.

Reviewer: These seems like a set of steps. Let's put them in an ordered list. Perhaps:
```suggestion
With this example, the following actions occur:
1. The code binds to the `input` event of the `<input>` element, which allows the code to listen for changes.
2. When the user makes changes, the component raises the `input` event.
3. The binding executes the statement within a context that includes the DOM event object, `$event`.
4. Angular retrieves the changed text by following the path `$event.target.value` and updates the `name` property.
```
One reason for this suggestion: a list can break up a block of text, making it easier for users to read.

If the event belongs to a directive or component, `$event` has the shape that the directive or component produces.

## Custom events with `EventEmitter`

> Directives typically raise custom events with an Angular [EventEmitter](api/core/EventEmitter).

Reviewer: Maybe link to our content on directives?

The directive creates an `EventEmitter` and exposes it as a property.
The directive then calls `EventEmitter.emit(data)` to emit an event, passing in message data, which can be anything.

> Parent directives listen for the event by binding to this property and accessing the data through the `$event` object.

Reviewer: As with a comment I made earlier: Can we rewrite this as an ordered list?

Consider an `ItemDetailComponent` that presents item information and responds to user actions.
Although the `ItemDetailComponent` has a delete button, it doesn't contain the functionality to delete the hero.
It can only raise an event reporting the user's delete request.

> In the following example, when the user clicks the delete button, the component raises the click event, but doesn't actually delete the item.

Reviewer: I think we can remove this sentence. It basically repeats the preceding one.

<code-example path="event-binding/src/app/item-detail/item-detail.component.html" header="src/app/item-detail/item-detail.component.html (template)" region="line-through"></code-example>

The component defines a `deleteRequest` property that returns an `EventEmitter`.
When the user clicks **Delete**, the component invokes the `delete()` method, telling the `EventEmitter` to emit an `Item` object.

<code-example path="event-binding/src/app/item-detail/item-detail.component.ts" header="src/app/item-detail/item-detail.component.ts (deleteRequest)" region="deleteRequest"></code-example>

The hosting parent component binds to the `deleteRequest` event of the `ItemDetailComponent` as follows.

<code-example path="event-binding/src/app/app.component.html" header="src/app/app.component.html (event-binding-to-component)" region="event-binding-to-component"></code-example>

When the `deleteRequest` event fires, Angular calls the parent component's `deleteItem()` method with the item.

> ### How Angular determines the target event

Reviewer: ```suggestion
### Determining an event target
```

> Angular looks first to see if the name of the target event matches an event property of a known directive.

Reviewer: ```suggestion
To determine an event target, Angular checks if the name of the target event matches an event property of a known directive.
```

> Angular looks first to see if the name of the target event matches an event property of a known directive.

Reviewer 2: ```suggestion
Angular looks first to see if the name of the target event matches an event property of a known component or directive.
```
I might make this a bit more explicit. Even though `Component` extends `Directive` (and thus a component _is_ a directive), only referring to "directive" can sometimes confuse people because it's ambiguous.

Author: Incorporated into rewrite of this sentence which now at line 30.

In the following example, Angular checks to see if `myClick` is an event on the custom `ClickDirective`.

<code-example path="event-binding/src/app/app.component.html" region="custom-directive" header="src/app/app.component.html"></code-example>

> <div class="alert is-helpful">
> 
> If the target event name fails to match an element event or an output property of a known directive, Angular reports an "unknown directive" error.
> 
> </div>

Reviewer: ```suggestion
If the target event name fails to match an element event or an output property of a known directive, Angular reports an `unknown directive` error.
```
A couple of thoughts:
This section is really just saying: "Angular checks to see if your template statement matches an actual function. If it doesn't, it throws an error.

The content in the alert box is essential to the section. In that situation, we shouldn't put it in a note box, because people might skip it. :)

Author: Updated and removed alert styling:

>If the target event name, `myClick` fails to match an element event or an output property of `ClickDirective`, Angular reports an "unknown directive" error.
